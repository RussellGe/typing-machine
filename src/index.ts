import type { Diff } from "diff-match-patch";
import { diff_match_patch as DMP } from "diff-match-patch";

export interface InsertPatch {
  type: "insert";
  from: number;
  text: string;
}
export interface RemovalPatch {
  type: "removal";
  from: number;
  length: number;
}

export type Patch = InsertPatch | RemovalPatch;

export function diff(a: string, b: string): Diff[] {
  const differ = new DMP();
  const delta = differ.diff_main(a, b);
  differ.diff_cleanupSemantic(delta);
  return delta;
}

export function calculatePatch(diffs: Diff[]): Patch[] {
  const patches: Patch[] = [];
  let index = 0;
  for (const change of diffs) {
    if (change[0] === 0) {
      index += change[1].length;
      continue;
    }
    if (change[0] === -1) {
      patches.push({
        type: "removal",
        from: index + change[1].length,
        length: change[1].length,
      });
      continue;
    }
    if (change[0] === 1) {
      patches.push({
        type: "insert",
        from: index,
        text: change[1],
      });
      index += change[1].length;
      continue;
    }
  }
  return patches;
}

export function applyPatches(input: string, patches: Patch[]) {
  let output = input;
  for (const patch of patches) {
    if (patch.type === "insert") {
      output =
        output.slice(0, patch.from) + patch.text + output.slice(patch.from);
    }
    if (patch.type === "removal") {
      output =
        output.slice(0, patch.from - patch.length) + output.slice(patch.from);
    }
  }
  return output;
}