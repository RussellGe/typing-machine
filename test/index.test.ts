import { describe, expect, it } from "vitest";
import { one } from "../src";
import type { Diff } from "diff-match-patch";
import { diff_match_patch as DMP } from "diff-match-patch";

const input = `import { describe, expect, it } from 'vitest'
import { one } from '../src'

describe('should', () => {
  it('exported', () => {
    expect(1).toEqual(1)
  })
})
`;

const output = `import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
`;

interface InsertPatch {
  type: "insert"
  from: number
  text: string
}
interface RemovalPatch {
  type: "removal"
  from: number
  length: number
}

type Patch = InsertPatch | RemovalPatch

function diff(a: string, b: string): Diff[] {
  const differ = new DMP();
  const delta = differ.diff_main(a, b);
  differ.diff_cleanupSemantic(delta);
  return delta;
}

function calculatePatch(diffs: Diff[]): Patch[] {
  const patches: Patch[] = []
  let index = 0
  for(const change of diffs) {
    if(change[0] === 0) {
      index += change[1].length
      continue
    }
    if(change[0] === -1) {
      patches.push({
        type: 'removal',
        from: index + change[1].length,
        length: change[1].length
      })
      continue
    }
    if(change[0] === 1) {
      patches.push({
        type: "insert",
        from: index,
        text: change[1]
      })
      index += change[1].length
      continue
    }
  }
  return patches
}

function applyPatches(input: string, patches: Patch[]) {
  let output = input
  for(const patch of patches) {
    if(patch.type === 'insert') {
      output = output.slice(0, patch.from) + patch.text + output.slice(patch.from)
    }
    if(patch.type === 'removal') {
      output = output.slice(0, patch.from - patch.length) + output.slice(patch.from)
    }
  }
  return output
}

describe("should", () => {
  it("exported", () => {
    const delta = diff(input, output)
    expect(delta).toMatchSnapshot('delta'); 
    const patches = calculatePatch(delta)
    expect(patches).toMatchSnapshot('patches');
    const outputs = applyPatches(input, patches)
    expect(outputs).toMatchSnapshot('outputs');
  });
});
