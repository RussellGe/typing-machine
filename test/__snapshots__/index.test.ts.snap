// Vitest Snapshot v1

exports[`should > exported > delta 1`] = `
[
  [
    0,
    "import { describe, expect, it } from 'vitest'",
  ],
  [
    -1,
    "
import { one } from '../src'",
  ],
  [
    0,
    "

describe('should', () => {
  it('",
  ],
  [
    -1,
    "exported",
  ],
  [
    1,
    "one",
  ],
  [
    0,
    "', () => {
    expect(",
  ],
  [
    -1,
    "1",
  ],
  [
    1,
    "one).toEqual(1)
    expect(two",
  ],
  [
    0,
    ").toEqual(",
  ],
  [
    -1,
    "1",
  ],
  [
    1,
    "2",
  ],
  [
    0,
    ")
  })
})
",
  ],
]
`;

exports[`should > exported > outputs 1`] = `
"import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
"
`;

exports[`should > exported > patches 1`] = `
[
  {
    "from": 74,
    "length": 29,
    "type": "removal",
  },
  {
    "from": 88,
    "length": 8,
    "type": "removal",
  },
  {
    "from": 80,
    "text": "one",
    "type": "insert",
  },
  {
    "from": 106,
    "length": 1,
    "type": "removal",
  },
  {
    "from": 105,
    "text": "one).toEqual(1)
    expect(two",
    "type": "insert",
  },
  {
    "from": 146,
    "length": 1,
    "type": "removal",
  },
  {
    "from": 145,
    "text": "2",
    "type": "insert",
  },
]
`;
